regression in `wasm32` support for latest releases of `spl-token-2022`

### Description

The most recent versions of this library don't support wasm32. Running the following leads to an error.

```bash
cargo build -p spl-token-2022 --target wasm32-unknown-unknown
```

This is the error. 

```bash
Compiling proc-macro2 v1.0.94
   Compiling unicode-ident v1.0.18
   Compiling version_check v0.9.5
   Compiling typenum v1.18.0
   Compiling cfg-if v1.0.0
   Compiling wasm-bindgen-shared v0.2.100
   Compiling once_cell v1.20.3
   Compiling bumpalo v3.17.0
   Compiling log v0.4.26
   Compiling rustversion v1.0.20
   Compiling wasm-bindgen v0.2.100
   Compiling autocfg v1.4.0
   Compiling subtle v2.6.1
   Compiling serde v1.0.219
   Compiling byteorder v1.5.0
   Compiling hashbrown v0.15.2
   Compiling equivalent v1.0.2
   Compiling winnow v0.7.3
   Compiling toml_datetime v0.6.8
   Compiling cfg_aliases v0.2.1
   Compiling parking_lot_core v0.9.10
   Compiling generic-array v0.14.7
   Compiling borsh v1.5.5
   Compiling lock_api v0.4.12
   Compiling num-traits v0.2.19
   Compiling indexmap v2.7.1
   Compiling semver v1.0.26
   Compiling smallvec v1.14.0
   Compiling syn v1.0.109
   Compiling scopeguard v1.2.0
   Compiling quote v1.0.39
   Compiling syn v2.0.99
   Compiling toml_edit v0.22.24
   Compiling ahash v0.8.11
   Compiling bs58 v0.5.1
   Compiling solana-sanitize v2.2.1
   Compiling five8_core v0.1.1
   Compiling five8_const v0.1.3
   Compiling solana-msg v2.2.1
   Compiling block-buffer v0.10.4
   Compiling lazy_static v1.5.0
   Compiling thiserror v2.0.12
   Compiling parking_lot v0.12.3
   Compiling opaque-debug v0.3.1
   Compiling keccak v0.1.5
   Compiling solana-atomic-u64 v2.2.1
   Compiling crunchy v0.2.3
   Compiling proc-macro-crate v3.3.0
   Compiling rustc_version v0.4.1
   Compiling inout v0.1.4
   Compiling base64 v0.22.1
   Compiling libc v0.2.170
   Compiling curve25519-dalek v4.1.3
   Compiling thiserror v1.0.69
   Compiling percent-encoding v2.3.1
   Compiling fnv v1.0.7
   Compiling solana-decode-error v2.2.1
   Compiling solana-program-memory v2.2.1
   Compiling uriparse v0.6.4
   Compiling qstring v0.7.2
   Compiling serde_json v1.0.140
   Compiling derivation-path v0.2.0
   Compiling digest v0.9.0
   Compiling crypto-common v0.1.6
   Compiling solana-signature v2.2.1
   Compiling either v1.15.0
   Compiling getrandom v0.1.16
   Compiling ryu v1.0.20
   Compiling memchr v2.7.4
   Compiling shlex v1.3.0
   Compiling itoa v1.0.15
   Compiling wasm-bindgen-backend v0.2.100
   Compiling solana-derivation-path v2.2.1
   Compiling libsecp256k1-core v0.2.2
   Compiling borsh-derive-internal v0.10.4
   Compiling borsh-schema-derive-internal v0.10.4
   Compiling solana-seed-derivable v2.2.1
   Compiling cc v1.2.16
   Compiling cpufeatures v0.2.17
   Compiling itertools v0.12.1
   Compiling digest v0.10.7
   Compiling arrayref v0.3.9
   Compiling solana-program-option v2.2.1
   Compiling feature-probe v0.1.1
   Compiling sha2 v0.10.8
   Compiling libsecp256k1-gen-ecmult v0.2.1
   Compiling libsecp256k1-gen-genmult v0.2.1
   Compiling rand_core v0.5.1
   Compiling bv v0.11.1
   Compiling arrayvec v0.7.6
   Compiling constant_time_eq v0.3.1
   Compiling bitflags v2.9.0
   Compiling libsecp256k1 v0.6.0
   Compiling block-buffer v0.9.0
   Compiling wasm-bindgen-macro-support v0.2.100
   Compiling toml v0.5.11
   Compiling sha2 v0.9.9
   Compiling num-integer v0.1.46
   Compiling memoffset v0.9.1
   Compiling base64 v0.12.3
   Compiling num-bigint v0.4.6
   Compiling solana-native-token v2.2.1
   Compiling proc-macro-crate v0.1.5
   Compiling solana-security-txt v1.1.1
   Compiling blake3 v1.6.1
   Compiling serde_derive v1.0.219
   Compiling zerocopy-derive v0.7.35
   Compiling borsh-derive v1.5.5
   Compiling bytemuck_derive v1.8.1
   Compiling zeroize_derive v1.4.2
   Compiling thiserror-impl v2.0.12
   Compiling solana-sdk-macro v2.2.1
   Compiling num-derive v0.4.2
   Compiling wasm-bindgen-macro v0.2.100
   Compiling borsh-derive v0.10.4
   Compiling thiserror-impl v1.0.69
   Compiling solana-big-mod-exp v2.2.1
   Compiling spl-program-error-derive v0.4.1
   Compiling num_enum_derive v0.7.3
   Compiling zeroize v1.8.1
   Compiling zerocopy v0.7.35
   Compiling bytemuck v1.22.0
   Compiling js-sys v0.3.77
   Compiling console_error_panic_hook v0.1.7
   Compiling num_enum v0.7.3
   Compiling ppv-lite86 v0.2.20
   Compiling hashbrown v0.13.2
   Compiling rand_chacha v0.2.2
   Compiling rand v0.7.3
   Compiling spl-discriminator-syn v0.2.0
   Compiling borsh v0.10.4
   Compiling spl-discriminator-derive v0.2.0
   Compiling solana-borsh v2.2.1
   Compiling bincode v1.3.3
   Compiling solana-fee-calculator v2.2.1
   Compiling serde_bytes v0.11.16
   Compiling solana-short-vec v2.2.1
   Compiling solana-serde-varint v2.2.1
   Compiling getrandom v0.2.15
   Compiling solana-hash v2.2.1
   Compiling web-sys v0.3.77
   Compiling rand_core v0.6.4
   Compiling rand_chacha v0.3.1
   Compiling merlin v3.0.0
   Compiling cipher v0.4.4
   Compiling universal-hash v0.5.1
   Compiling aead v0.5.2
   Compiling rand v0.8.5
   Compiling polyval v0.6.2
   Compiling sha3 v0.10.8
   Compiling hmac v0.12.1
   Compiling pbkdf2 v0.11.0
   Compiling aes v0.8.4
   Compiling ctr v0.9.2
   Compiling solana-sha256-hasher v2.2.1
   Compiling solana-seed-phrase v2.2.1
   Compiling solana-keccak-hasher v2.2.1
   Compiling aes-gcm-siv v0.11.1
   Compiling solana-blake3-hasher v2.2.1
   Compiling console_log v0.2.2
   Compiling solana-pubkey v2.2.1
   Compiling solana-curve25519 v2.2.1
   Compiling solana-instruction v2.2.1
   Compiling solana-sdk-ids v2.2.1
   Compiling solana-nonce v2.2.1
   Compiling solana-sysvar-id v2.2.1
   Compiling solana-clock v2.2.1
   Compiling solana-slot-hashes v2.2.1
   Compiling solana-rent v2.2.1
   Compiling solana-epoch-rewards v2.2.1
   Compiling solana-epoch-schedule v2.2.1
   Compiling solana-slot-history v2.2.1
   Compiling solana-last-restart-slot v2.2.1
   Compiling solana-program-error v2.2.1
   Compiling solana-transaction-error v2.2.1
   Compiling solana-system-interface v1.0.0
   Compiling solana-serialize-utils v2.2.1
   Compiling solana-bincode v2.2.1
   Compiling solana-account-info v2.2.1
   Compiling solana-signer v2.2.1
   Compiling spl-discriminator v0.4.1
   Compiling solana-program-pack v2.2.1
   Compiling solana-cpi v2.2.1
   Compiling solana-program-entrypoint v2.2.1
   Compiling solana-instructions-sysvar v2.2.1
   Compiling solana-address-lookup-table-interface v2.2.2
   Compiling solana-account v2.2.1
   Compiling solana-secp256k1-recover v2.2.1
   Compiling solana-stable-layout v2.2.1
   Compiling solana-loader-v2-interface v2.2.1
   Compiling spl-memo v6.0.0
   Compiling solana-stake-interface v1.2.1
   Compiling solana-message v2.2.1
   Compiling solana-loader-v4-interface v2.2.1
   Compiling solana-vote-interface v2.2.1
   Compiling solana-feature-gate-interface v2.2.1
   Compiling solana-loader-v3-interface v3.0.0
   Compiling solana-example-mocks v2.2.1
   Compiling solana-sysvar v2.2.1
   Compiling solana-zk-sdk v2.2.1
   Compiling solana-program v2.2.1
   Compiling spl-program-error v0.6.0
   Compiling spl-token v7.0.0
   Compiling spl-pod v0.5.1
   Compiling spl-token-confidential-transfer-proof-generation v0.3.0 (/Users/ifiokjr/Developer/os/token-2022/confidential-transfer/proof-generation)
   Compiling spl-token-confidential-transfer-ciphertext-arithmetic v0.2.1 (/Users/ifiokjr/Developer/os/token-2022/confidential-transfer/ciphertext-arithmetic)
error[E0308]: arguments to this function are incorrect
   --> confidential-transfer/proof-generation/src/burn.rs:107:42
    |
107 |     let ciphertext_validity_proof_data = BatchedGroupedCiphertext3HandlesValidityProofData::new(
    |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: expected `&GroupedElGamalCiphertext3Handles`, found `&GroupedElGamalCiphertext<3>`
   --> confidential-transfer/proof-generation/src/burn.rs:111:9
    |
111 |         &burn_amount_ciphertext_lo.0,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext3Handles`
               found reference `&GroupedElGamalCiphertext<3>`
note: expected `&GroupedElGamalCiphertext3Handles`, found `&GroupedElGamalCiphertext<3>`
   --> confidential-transfer/proof-generation/src/burn.rs:112:9
    |
112 |         &burn_amount_ciphertext_hi.0,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext3Handles`
               found reference `&GroupedElGamalCiphertext<3>`
note: associated function defined here
   --> /Users/ifiokjr/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/solana-zk-sdk-2.2.1/src/zk_elgamal_proof_program/proof_data/batched_grouped_ciphertext_validity/handles_3.rs:120:12
    |
120 |     pub fn new(
    |            ^^^

error[E0308]: arguments to this function are incorrect
   --> confidential-transfer/proof-generation/src/mint.rs:100:42
    |
100 |     let ciphertext_validity_proof_data = BatchedGroupedCiphertext3HandlesValidityProofData::new(
    |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: expected `&GroupedElGamalCiphertext3Handles`, found `&GroupedElGamalCiphertext<3>`
   --> confidential-transfer/proof-generation/src/mint.rs:104:9
    |
104 |         &mint_amount_grouped_ciphertext_lo.0,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext3Handles`
               found reference `&GroupedElGamalCiphertext<3>`
note: expected `&GroupedElGamalCiphertext3Handles`, found `&GroupedElGamalCiphertext<3>`
   --> confidential-transfer/proof-generation/src/mint.rs:105:9
    |
105 |         &mint_amount_grouped_ciphertext_hi.0,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext3Handles`
               found reference `&GroupedElGamalCiphertext<3>`
note: associated function defined here
   --> /Users/ifiokjr/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/solana-zk-sdk-2.2.1/src/zk_elgamal_proof_program/proof_data/batched_grouped_ciphertext_validity/handles_3.rs:120:12
    |
120 |     pub fn new(
    |            ^^^

error[E0308]: arguments to this function are incorrect
   --> confidential-transfer/proof-generation/src/transfer.rs:111:42
    |
111 |     let ciphertext_validity_proof_data = BatchedGroupedCiphertext3HandlesValidityProofData::new(
    |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: expected `&GroupedElGamalCiphertext3Handles`, found `&GroupedElGamalCiphertext<3>`
   --> confidential-transfer/proof-generation/src/transfer.rs:115:9
    |
115 |         &transfer_amount_grouped_ciphertext_lo.0,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext3Handles`
               found reference `&GroupedElGamalCiphertext<3>`
note: expected `&GroupedElGamalCiphertext3Handles`, found `&GroupedElGamalCiphertext<3>`
   --> confidential-transfer/proof-generation/src/transfer.rs:116:9
    |
116 |         &transfer_amount_grouped_ciphertext_hi.0,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext3Handles`
               found reference `&GroupedElGamalCiphertext<3>`
note: associated function defined here
   --> /Users/ifiokjr/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/solana-zk-sdk-2.2.1/src/zk_elgamal_proof_program/proof_data/batched_grouped_ciphertext_validity/handles_3.rs:120:12
    |
120 |     pub fn new(
    |            ^^^

error[E0308]: arguments to this function are incorrect
   --> confidential-transfer/proof-generation/src/transfer_with_fee.rs:129:9
    |
129 |         BatchedGroupedCiphertext3HandlesValidityProofData::new(
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: expected `&GroupedElGamalCiphertext3Handles`, found `&GroupedElGamalCiphertext<3>`
   --> confidential-transfer/proof-generation/src/transfer_with_fee.rs:133:13
    |
133 |             &transfer_amount_grouped_ciphertext_lo.0,
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext3Handles`
               found reference `&GroupedElGamalCiphertext<3>`
note: expected `&GroupedElGamalCiphertext3Handles`, found `&GroupedElGamalCiphertext<3>`
   --> confidential-transfer/proof-generation/src/transfer_with_fee.rs:134:13
    |
134 |             &transfer_amount_grouped_ciphertext_hi.0,
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext3Handles`
               found reference `&GroupedElGamalCiphertext<3>`
note: associated function defined here
   --> /Users/ifiokjr/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/solana-zk-sdk-2.2.1/src/zk_elgamal_proof_program/proof_data/batched_grouped_ciphertext_validity/handles_3.rs:120:12
    |
120 |     pub fn new(
    |            ^^^

error[E0308]: arguments to this function are incorrect
   --> confidential-transfer/proof-generation/src/transfer_with_fee.rs:255:9
    |
255 |         BatchedGroupedCiphertext2HandlesValidityProofData::new(
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: expected `&GroupedElGamalCiphertext2Handles`, found `&GroupedElGamalCiphertext<2>`
   --> confidential-transfer/proof-generation/src/transfer_with_fee.rs:258:13
    |
258 |             &fee_destination_withdraw_withheld_authority_ciphertext_lo,
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext2Handles`
               found reference `&GroupedElGamalCiphertext<2>`
note: expected `&GroupedElGamalCiphertext2Handles`, found `&GroupedElGamalCiphertext<2>`
   --> confidential-transfer/proof-generation/src/transfer_with_fee.rs:259:13
    |
259 |             &fee_destination_withdraw_withheld_authority_ciphertext_hi,
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: expected reference `&GroupedElGamalCiphertext2Handles`
               found reference `&GroupedElGamalCiphertext<2>`
note: associated function defined here
   --> /Users/ifiokjr/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/solana-zk-sdk-2.2.1/src/zk_elgamal_proof_program/proof_data/batched_grouped_ciphertext_validity/handles_2.rs:114:12
    |
114 |     pub fn new(
    |            ^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `spl-token-confidential-transfer-proof-generation` (lib) due to 5 previous errors
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -286,6 +286,9 @@ jobs:
       - name: Build Token-2022
         run: pnpm programs:build
 
+      - name: Build Token-2022 Wasm
+        run: pnpm wasm:build
+
       - name: Build ElGamal Registry
         run: pnpm confidential-transfer:elgamal-registry:build
 

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -287,7 +287,7 @@ jobs:
         run: pnpm programs:build
 
       - name: Build Token-2022 Wasm
-        run: pnpm wasm:build
+        run: pnpm programs:build-wasm
 
       - name: Build ElGamal Registry
         run: pnpm confidential-transfer:elgamal-registry:build

--- a/confidential-transfer/proof-generation/src/burn.rs
+++ b/confidential-transfer/proof-generation/src/burn.rs
@@ -1,3 +1,5 @@
+#[cfg(target_arch = "wasm32")]
+use solana_zk_sdk::encryption::grouped_elgamal::GroupedElGamalCiphertext3Handles;
 use {
     crate::{
         encryption::BurnAmountCiphertext, errors::TokenProofGenerationError,
@@ -53,6 +55,16 @@ pub fn burn_split_proof_data(
         supply_elgamal_pubkey,
         auditor_elgamal_pubkey,
     );
+    #[cfg(not(target_arch = "wasm32"))]
+    let grouped_ciphertext_lo = burn_amount_ciphertext_lo.0;
+    #[cfg(target_arch = "wasm32")]
+    let grouped_ciphertext_lo = GroupedElGamalCiphertext3Handles::encryption_with_u64(
+        source_elgamal_keypair.pubkey(),
+        supply_elgamal_pubkey,
+        auditor_elgamal_pubkey,
+        burn_amount_lo,
+        &burn_amount_opening_lo,
+    );
 
     let (burn_amount_ciphertext_hi, burn_amount_opening_hi) = BurnAmountCiphertext::new(
         burn_amount_hi,
@@ -60,7 +72,16 @@ pub fn burn_split_proof_data(
         supply_elgamal_pubkey,
         auditor_elgamal_pubkey,
     );
-
+    #[cfg(not(target_arch = "wasm32"))]
+    let grouped_ciphertext_hi = burn_amount_ciphertext_hi.0;
+    #[cfg(target_arch = "wasm32")]
+    let grouped_ciphertext_hi = GroupedElGamalCiphertext3Handles::encryption_with_u64(
+        source_elgamal_keypair.pubkey(),
+        supply_elgamal_pubkey,
+        auditor_elgamal_pubkey,
+        burn_amount_hi,
+        &burn_amount_opening_hi,
+    );
     // decrypt the current available balance at the source
     let current_decrypted_available_balance = current_decryptable_available_balance
         .decrypt(source_aes_key)
@@ -108,8 +129,8 @@ pub fn burn_split_proof_data(
         source_elgamal_keypair.pubkey(),
         supply_elgamal_pubkey,
         auditor_elgamal_pubkey,
-        &burn_amount_ciphertext_lo.0,
-        &burn_amount_ciphertext_hi.0,
+        &grouped_ciphertext_lo,
+        &grouped_ciphertext_hi,
         burn_amount_lo,
         burn_amount_hi,
         &burn_amount_opening_lo,

--- a/confidential-transfer/proof-generation/src/mint.rs
+++ b/confidential-transfer/proof-generation/src/mint.rs
@@ -1,3 +1,5 @@
+#[cfg(target_arch = "wasm32")]
+use solana_zk_sdk::encryption::grouped_elgamal::GroupedElGamalCiphertext3Handles;
 use {
     crate::{
         encryption::MintAmountCiphertext, errors::TokenProofGenerationError,
@@ -52,12 +54,32 @@ pub fn mint_split_proof_data(
         supply_elgamal_keypair.pubkey(),
         auditor_elgamal_pubkey,
     );
+    #[cfg(not(target_arch = "wasm32"))]
+    let grouped_ciphertext_lo = mint_amount_grouped_ciphertext_lo.0;
+    #[cfg(target_arch = "wasm32")]
+    let grouped_ciphertext_lo = GroupedElGamalCiphertext3Handles::encryption_with_u64(
+        destination_elgamal_pubkey,
+        supply_elgamal_keypair.pubkey(),
+        auditor_elgamal_pubkey,
+        mint_amount_lo,
+        &mint_amount_opening_lo,
+    );
 
     let (mint_amount_grouped_ciphertext_hi, mint_amount_opening_hi) = MintAmountCiphertext::new(
         mint_amount_hi,
         destination_elgamal_pubkey,
         supply_elgamal_keypair.pubkey(),
         auditor_elgamal_pubkey,
+    );
+    #[cfg(not(target_arch = "wasm32"))]
+    let grouped_ciphertext_hi = mint_amount_grouped_ciphertext_hi.0;
+    #[cfg(target_arch = "wasm32")]
+    let grouped_ciphertext_hi = GroupedElGamalCiphertext3Handles::encryption_with_u64(
+        destination_elgamal_pubkey,
+        supply_elgamal_keypair.pubkey(),
+        auditor_elgamal_pubkey,
+        mint_amount_hi,
+        &mint_amount_opening_hi,
     );
 
     // compute the new supply ciphertext
@@ -101,8 +123,8 @@ pub fn mint_split_proof_data(
         destination_elgamal_pubkey,
         supply_elgamal_keypair.pubkey(),
         auditor_elgamal_pubkey,
-        &mint_amount_grouped_ciphertext_lo.0,
-        &mint_amount_grouped_ciphertext_hi.0,
+        &grouped_ciphertext_lo,
+        &grouped_ciphertext_hi,
         mint_amount_lo,
         mint_amount_hi,
         &mint_amount_opening_lo,

--- a/confidential-transfer/proof-generation/src/transfer.rs
+++ b/confidential-transfer/proof-generation/src/transfer.rs
@@ -1,3 +1,5 @@
+#[cfg(target_arch = "wasm32")]
+use solana_zk_sdk::encryption::grouped_elgamal::GroupedElGamalCiphertext3Handles;
 use {
     crate::{
         encryption::TransferAmountCiphertext, errors::TokenProofGenerationError,
@@ -55,6 +57,16 @@ pub fn transfer_split_proof_data(
             destination_elgamal_pubkey,
             auditor_elgamal_pubkey,
         );
+    #[cfg(not(target_arch = "wasm32"))]
+    let grouped_ciphertext_lo = transfer_amount_grouped_ciphertext_lo.0;
+    #[cfg(target_arch = "wasm32")]
+    let grouped_ciphertext_lo = GroupedElGamalCiphertext3Handles::encryption_with_u64(
+        source_elgamal_keypair.pubkey(),
+        destination_elgamal_pubkey,
+        auditor_elgamal_pubkey,
+        transfer_amount_lo,
+        &transfer_amount_opening_lo,
+    );
 
     let (transfer_amount_grouped_ciphertext_hi, transfer_amount_opening_hi) =
         TransferAmountCiphertext::new(
@@ -63,6 +75,16 @@ pub fn transfer_split_proof_data(
             destination_elgamal_pubkey,
             auditor_elgamal_pubkey,
         );
+    #[cfg(not(target_arch = "wasm32"))]
+    let grouped_ciphertext_hi = transfer_amount_grouped_ciphertext_hi.0;
+    #[cfg(target_arch = "wasm32")]
+    let grouped_ciphertext_hi = GroupedElGamalCiphertext3Handles::encryption_with_u64(
+        source_elgamal_keypair.pubkey(),
+        destination_elgamal_pubkey,
+        auditor_elgamal_pubkey,
+        transfer_amount_hi,
+        &transfer_amount_opening_hi,
+    );
 
     // Decrypt the current available balance at the source
     let current_decrypted_available_balance = current_decryptable_available_balance
@@ -112,8 +134,8 @@ pub fn transfer_split_proof_data(
         source_elgamal_keypair.pubkey(),
         destination_elgamal_pubkey,
         auditor_elgamal_pubkey,
-        &transfer_amount_grouped_ciphertext_lo.0,
-        &transfer_amount_grouped_ciphertext_hi.0,
+        &grouped_ciphertext_lo,
+        &grouped_ciphertext_hi,
         transfer_amount_lo,
         transfer_amount_hi,
         &transfer_amount_opening_lo,

--- a/confidential-transfer/proof-generation/src/transfer_with_fee.rs
+++ b/confidential-transfer/proof-generation/src/transfer_with_fee.rs
@@ -1,3 +1,7 @@
+#[cfg(target_arch = "wasm32")]
+use solana_zk_sdk::encryption::grouped_elgamal::{GroupedElGamalCiphertext2Handles, GroupedElGamalCiphertext3Handles};
+#[cfg(not(target_arch = "wasm32"))]
+use solana_zk_sdk::encryption::grouped_elgamal::GroupedElGamal;
 use {
     crate::{
         encryption::{FeeCiphertext, TransferAmountCiphertext},
@@ -9,10 +13,7 @@ use {
     curve25519_dalek::scalar::Scalar,
     solana_zk_sdk::{
         encryption::{
-            auth_encryption::{AeCiphertext, AeKey},
-            elgamal::{ElGamalCiphertext, ElGamalKeypair, ElGamalPubkey},
-            grouped_elgamal::GroupedElGamal,
-            pedersen::{Pedersen, PedersenCommitment, PedersenOpening},
+            auth_encryption::{AeCiphertext, AeKey}, elgamal::{ElGamalCiphertext, ElGamalKeypair, ElGamalPubkey}, pedersen::{Pedersen, PedersenCommitment, PedersenOpening}
         },
         zk_elgamal_proof_program::proof_data::{
             BatchedGroupedCiphertext2HandlesValidityProofData,
@@ -71,6 +72,16 @@ pub fn transfer_with_fee_split_proof_data(
             destination_elgamal_pubkey,
             auditor_elgamal_pubkey,
         );
+    #[cfg(not(target_arch = "wasm32"))]
+    let grouped_ciphertext_lo = transfer_amount_grouped_ciphertext_lo.0;
+    #[cfg(target_arch = "wasm32")]
+    let grouped_ciphertext_lo = GroupedElGamalCiphertext3Handles::encryption_with_u64(
+        source_elgamal_keypair.pubkey(),
+        destination_elgamal_pubkey,
+        auditor_elgamal_pubkey,
+        transfer_amount_lo,
+        &transfer_amount_opening_lo,
+    );
 
     let (transfer_amount_grouped_ciphertext_hi, transfer_amount_opening_hi) =
         TransferAmountCiphertext::new(
@@ -79,6 +90,16 @@ pub fn transfer_with_fee_split_proof_data(
             destination_elgamal_pubkey,
             auditor_elgamal_pubkey,
         );
+    #[cfg(not(target_arch = "wasm32"))]
+    let grouped_ciphertext_hi = transfer_amount_grouped_ciphertext_hi.0;
+    #[cfg(target_arch = "wasm32")]
+    let grouped_ciphertext_hi = GroupedElGamalCiphertext3Handles::encryption_with_u64(
+        source_elgamal_keypair.pubkey(),
+        destination_elgamal_pubkey,
+        auditor_elgamal_pubkey,
+        transfer_amount_hi,
+        &transfer_amount_opening_hi,
+    );
 
     // Decrypt the current available balance at the source
     let current_decrypted_available_balance = current_decryptable_available_balance
@@ -130,8 +151,8 @@ pub fn transfer_with_fee_split_proof_data(
             source_elgamal_keypair.pubkey(),
             destination_elgamal_pubkey,
             auditor_elgamal_pubkey,
-            &transfer_amount_grouped_ciphertext_lo.0,
-            &transfer_amount_grouped_ciphertext_hi.0,
+            &grouped_ciphertext_lo,
+            &grouped_ciphertext_hi,
             transfer_amount_lo,
             transfer_amount_hi,
             &transfer_amount_opening_lo,
@@ -233,6 +254,7 @@ pub fn transfer_with_fee_split_proof_data(
 
     // encrypt the fee amount under the destination and withdraw withheld authority
     // ElGamal public key
+    #[cfg(not(target_arch = "wasm32"))]
     let fee_destination_withdraw_withheld_authority_ciphertext_lo = GroupedElGamal::encrypt_with(
         [
             destination_elgamal_pubkey,
@@ -241,11 +263,26 @@ pub fn transfer_with_fee_split_proof_data(
         fee_amount_lo,
         &fee_opening_lo,
     );
-    let fee_destination_withdraw_withheld_authority_ciphertext_hi = GroupedElGamal::encrypt_with(
-        [
+    #[cfg(target_arch = "wasm32")]
+    let fee_destination_withdraw_withheld_authority_ciphertext_lo = GroupedElGamalCiphertext2Handles::encryption_with_u64(
+        destination_elgamal_pubkey,
+        withdraw_withheld_authority_elgamal_pubkey,
+        fee_amount_lo,
+        &fee_opening_lo
+    );
+
+    #[cfg(not(target_arch = "wasm32"))]
+    let fee_destination_withdraw_withheld_authority_ciphertext_hi = GroupedElGamal::encrypt_with([
             destination_elgamal_pubkey,
             withdraw_withheld_authority_elgamal_pubkey,
         ],
+        fee_amount_hi,
+        &fee_opening_hi,
+    );
+    #[cfg(target_arch = "wasm32")]
+    let fee_destination_withdraw_withheld_authority_ciphertext_hi = GroupedElGamalCiphertext2Handles::encryption_with_u64(
+        destination_elgamal_pubkey,
+        withdraw_withheld_authority_elgamal_pubkey,
         fee_amount_hi,
         &fee_opening_hi,
     );

--- a/confidential-transfer/proof-generation/src/transfer_with_fee.rs
+++ b/confidential-transfer/proof-generation/src/transfer_with_fee.rs
@@ -1,7 +1,9 @@
-#[cfg(target_arch = "wasm32")]
-use solana_zk_sdk::encryption::grouped_elgamal::{GroupedElGamalCiphertext2Handles, GroupedElGamalCiphertext3Handles};
 #[cfg(not(target_arch = "wasm32"))]
 use solana_zk_sdk::encryption::grouped_elgamal::GroupedElGamal;
+#[cfg(target_arch = "wasm32")]
+use solana_zk_sdk::encryption::grouped_elgamal::{
+    GroupedElGamalCiphertext2Handles, GroupedElGamalCiphertext3Handles,
+};
 use {
     crate::{
         encryption::{FeeCiphertext, TransferAmountCiphertext},
@@ -13,7 +15,9 @@ use {
     curve25519_dalek::scalar::Scalar,
     solana_zk_sdk::{
         encryption::{
-            auth_encryption::{AeCiphertext, AeKey}, elgamal::{ElGamalCiphertext, ElGamalKeypair, ElGamalPubkey}, pedersen::{Pedersen, PedersenCommitment, PedersenOpening}
+            auth_encryption::{AeCiphertext, AeKey},
+            elgamal::{ElGamalCiphertext, ElGamalKeypair, ElGamalPubkey},
+            pedersen::{Pedersen, PedersenCommitment, PedersenOpening},
         },
         zk_elgamal_proof_program::proof_data::{
             BatchedGroupedCiphertext2HandlesValidityProofData,
@@ -264,15 +268,17 @@ pub fn transfer_with_fee_split_proof_data(
         &fee_opening_lo,
     );
     #[cfg(target_arch = "wasm32")]
-    let fee_destination_withdraw_withheld_authority_ciphertext_lo = GroupedElGamalCiphertext2Handles::encryption_with_u64(
-        destination_elgamal_pubkey,
-        withdraw_withheld_authority_elgamal_pubkey,
-        fee_amount_lo,
-        &fee_opening_lo
-    );
+    let fee_destination_withdraw_withheld_authority_ciphertext_lo =
+        GroupedElGamalCiphertext2Handles::encryption_with_u64(
+            destination_elgamal_pubkey,
+            withdraw_withheld_authority_elgamal_pubkey,
+            fee_amount_lo,
+            &fee_opening_lo,
+        );
 
     #[cfg(not(target_arch = "wasm32"))]
-    let fee_destination_withdraw_withheld_authority_ciphertext_hi = GroupedElGamal::encrypt_with([
+    let fee_destination_withdraw_withheld_authority_ciphertext_hi = GroupedElGamal::encrypt_with(
+        [
             destination_elgamal_pubkey,
             withdraw_withheld_authority_elgamal_pubkey,
         ],
@@ -280,12 +286,13 @@ pub fn transfer_with_fee_split_proof_data(
         &fee_opening_hi,
     );
     #[cfg(target_arch = "wasm32")]
-    let fee_destination_withdraw_withheld_authority_ciphertext_hi = GroupedElGamalCiphertext2Handles::encryption_with_u64(
-        destination_elgamal_pubkey,
-        withdraw_withheld_authority_elgamal_pubkey,
-        fee_amount_hi,
-        &fee_opening_hi,
-    );
+    let fee_destination_withdraw_withheld_authority_ciphertext_hi =
+        GroupedElGamalCiphertext2Handles::encryption_with_u64(
+            destination_elgamal_pubkey,
+            withdraw_withheld_authority_elgamal_pubkey,
+            fee_amount_hi,
+            &fee_opening_hi,
+        );
 
     // generate fee ciphertext validity data
     let fee_ciphertext_validity_proof_data =

--- a/package.json
+++ b/package.json
@@ -1,8 +1,8 @@
 {
   "private": true,
   "scripts": {
-    "wasm:build": "cargo build -p spl-token-2022 --target=\"wasm32-unknown-unknown\"",
     "programs:build": "zx ./scripts/rust/build-sbf.mjs program",
+    "programs:build-wasm": "zx ./scripts/rust/build-wasm.mjs program",
     "programs:test": "zx ./scripts/rust/test-sbf.mjs program",
     "programs:format": "zx ./scripts/rust/format.mjs program",
     "programs:lint": "zx ./scripts/rust/lint.mjs program",

--- a/package.json
+++ b/package.json
@@ -1,6 +1,7 @@
 {
   "private": true,
   "scripts": {
+    "wasm:build": "cargo build -p spl-token-2022 --target=\"wasm32-unknown-unknown\"",
     "programs:build": "zx ./scripts/rust/build-sbf.mjs program",
     "programs:test": "zx ./scripts/rust/test-sbf.mjs program",
     "programs:format": "zx ./scripts/rust/format.mjs program",

--- a/program/src/extension/confidential_transfer_fee/processor.rs
+++ b/program/src/extension/confidential_transfer_fee/processor.rs
@@ -29,7 +29,6 @@ use {
         processor::Processor,
         solana_zk_sdk::encryption::pod::elgamal::PodElGamalPubkey,
     },
-    bytemuck::Zeroable,
     solana_account_info::{next_account_info, AccountInfo},
     solana_msg::msg,
     solana_program_error::{ProgramError, ProgramResult},
@@ -37,6 +36,9 @@ use {
     spl_pod::optional_keys::OptionalNonZeroPubkey,
     spl_token_confidential_transfer_proof_extraction::instruction::verify_and_extract_context,
 };
+
+#[cfg(not(target_arch = "wasm32"))]
+use bytemuck::Zeroable;
 
 /// Processes an [`InitializeConfidentialTransferFeeConfig`] instruction.
 fn process_initialize_confidential_transfer_fee_config(

--- a/rust-toolchain.toml
+++ b/rust-toolchain.toml
@@ -1,2 +1,3 @@
 [toolchain]
 channel = "1.84.1"
+targets = ["wasm32-unknown-unknown"]

--- a/scripts/rust/build-wasm.mjs
+++ b/scripts/rust/build-wasm.mjs
@@ -1,0 +1,10 @@
+#!/usr/bin/env zx
+import 'zx/globals';
+import {
+  cliArguments,
+  workingDirectory,
+} from '../utils.mjs';
+
+const [folder, ...args] = cliArguments();
+const manifestPath = path.join(workingDirectory, folder, 'Cargo.toml');
+await $`cargo build --target wasm32-unknown-unknown --manifest-path ${manifestPath} ${args}`;

